runtime/shared_common/CacheMap.cpp

```diff
@@ -7012,45 +7009,153 @@ SH_CacheMap::startupForStats(J9VMThread* currentThread, SH_OSCache * oscache, U_
 			retval = CC_STARTUP_FAILED;
 		}
 		goto done;
+	} else {
+		if (oscache->getLayer() > 0) {
+			startuprc = startupLowerLayerForStats(currentThread, ctrlDirName, groupPerm, oscache, lowerLayerList);
+			if (startuprc != CC_STARTUP_OK) {
+				if (startuprc == CC_STARTUP_CORRUPT) {
+					retval = CC_STARTUP_CORRUPT;
+				} else {
+					retval = CC_STARTUP_FAILED;
+				}
+				goto done;
+			}
+		}
 	}
 	setCacheAddressRangeArray();
 
-	if (oscache->getLayer() == 0) {
-		/* TODO: Should also read cache if layer > 0 */
-		/* populate the hashtables */
-		itemsRead = readCache(currentThread, _ccHead, -1, true);
+	ccToUse = _ccTail;
+	
+	do {
+		itemsRead = readCache(currentThread, ccToUse, -1, true);
 		if (CM_READ_CACHE_FAILED == itemsRead) {
 			retval = CC_STARTUP_FAILED;
-			goto done;
 		} else if (CM_CACHE_CORRUPT == itemsRead) {
 			retval = CC_STARTUP_CORRUPT;
-			goto done;
 		}
-	}
 
-#if defined(J9SHR_CACHELET_SUPPORT)
-	/* startup all cachelets to get stats */
-	while (cache) {
-		IDATA startupcachelet = startupCacheletForStats(currentThread, cache);
-		if ( CC_STARTUP_OK != startupcachelet ) {
-			if (startupcachelet == CC_STARTUP_CORRUPT) {
-				retval = CC_STARTUP_CORRUPT;
+		if (ccToUse != _ccHead) {
+			if (NULL == *lowerLayerList) {
+				*lowerLayerList = pool_new(sizeof(SH_OSCache_Info),  0, 0, 0, J9_GET_CALLSITE(), J9MEM_CATEGORY_CLASSES, POOL_FOR_PORT(_portlib));
+			}
+			if (NULL == *lowerLayerList) {
+				retval = CC_STARTUP_FAILED;
+				break;
+			}
+			(*lowerLayerList)->flags |= POOL_ALWAYS_KEEP_SORTED;
+			SH_OSCache_Info tempInfo;
+			if (-1 != ccToUse->getNonTopLayerCacheInfo(vm, ctrlDirName, groupPerm, &tempInfo)) {
+				if (CC_STARTUP_CORRUPT == retval) {
+					tempInfo.isCorrupt = 1;
+				}
+				if (!ccToUse->getJavacoreData(vm, &tempInfo.javacoreData)) {
+					retval = CC_STARTUP_FAILED;
+					break;
+				}
+				tempInfo.javacoreData.freeBytes = ccToUse->getFreeAvailableBytes();
+				tempInfo.isJavaCorePopulated = 1;
+				SH_OSCache_Info* newElement = (SH_OSCache_Info*) pool_newElement(*lowerLayerList);
+				if (NULL == newElement) {
+					retval = CC_STARTUP_FAILED;
+					break;
+				}
+				memcpy(newElement, &tempInfo, sizeof(SH_OSCache_Info));
 			} else {
 				retval = CC_STARTUP_FAILED;
+				break;
 			}
-			goto done;
 		}
-		cache = cache->getNext();
-	}
-#endif
+		ccToUse = ccToUse->getPrevious();
+	} while (NULL != ccToUse && CC_STARTUP_OK == retval);
 
 done:
 	if (retval != CC_STARTUP_OK) {
+		if (CC_STARTUP_FAILED == retval) {
```

Comment:
Why only for the CC_STARTUP_FAILED case? Shouldn't it also be cleared for CC_STARTUP_CORRUPT?

Reply:
We allow CC_STARTUP_CORRUPT and return lowerLayerList with SH_OSCache_Info.isCorrupt = 1.

Reply:
Is the pool guaranteed to be killed in that case?

Reply:
Then it is probably easier to call pool_kill() inside getAllCacheStatistics(). I have just pushed the change to always call pool_kill() in getAllCacheStatistics().